analytics.php

```diff
@@ -0,0 +1,231 @@
+    echo $OUTPUT->header();
+}
+$wheresql = "meta is not null and bigbluebuttonbnid = ? and log = ? and recordid = ? and userid is not null";
+$table->set_sql('id, meta, userid', "{bigbluebuttonbn_logs}", $wheresql, [
+    $bigbluebuttonbn->id,
+    BIGBLUEBUTTON_LOG_EVENT_SUMMARY,
+    $id
+]);
+$table->define_baseurl($pageurl);
+$meetingsummary = $DB->get_record('bigbluebuttonbn_logs', [
+    'userid' => null,
+    'recordid' => $id,
+    'log' => BIGBLUEBUTTON_LOG_EVENT_SUMMARY
+]);
+$output = "";
+
+// Prepare attendee data for Attention Box.
+$table->setup();
+$table->query_db(0); // No limit, fetch all rows.
+$table->pageable(false);
+$table->close_recordset();
+$attendees = $table->rawdata;
+
+if (isset($meetingsummary->meta)) {
+    $meetingsummary->meta = json_decode($meetingsummary->meta);
+}
+
+if (!$table->is_downloading()) {
+    $tables = "";
+    if (isset($meetingsummary->meta)) {
+        $output .= $OUTPUT->heading($meetingsummary->meta->meetingname, 3);
+        $output .= $OUTPUT->heading(
+            userdate(
+                bigbluebuttonbn_datetime_to_timestamp($meetingsummary->meta->start),
+                get_string('strftimedatetime', 'langconfig')
+            ),
+            6
+        );
+
+        // Overview Box.
+        $overviewtable = new html_table();
+        $overviewtable->head = [
+            get_string('view_analytics_heading_overview', 'bigbluebuttonbn'),
+            '',
+        ];
+        $overviewtable->attributes['class'] = 'w-auto admintable generaltable mr-4';
+        $durationhms = userdate($meetingsummary->meta->duration, "%H:%M:%S", 'UTC');
+        $htmlparams = ['class' => 'text-center'];
+        $data = [
+            [
+                get_string('view_analytics_duration', 'bigbluebuttonbn'),
+                \html_writer::tag('div', $durationhms, ['class' => 'text-right'])
+            ],
+            [
+                get_string('view_analytics_moderators', 'bigbluebuttonbn'),
+                \html_writer::tag('div', $meetingsummary->meta->moderators, $htmlparams),
+            ],
+            [
+                get_string('view_analytics_viewers', 'bigbluebuttonbn'),
+                \html_writer::tag('div', $meetingsummary->meta->viewers, $htmlparams),
+            ]
+        ];
+        $overviewtable->data = $data;
+        $overviewtablehtml = html_writer::table($overviewtable);
+        $tables .= $overviewtablehtml;
+    }
+
+    // Init data for Attention box.
+    $attentiontotals = [];
+
+    // Loop through and add +1 for every student who engages in the session, against the type of engagement (regardless of amount).
+    foreach ($attendees as &$attendee) {
+        $meta = json_decode($attendee->meta);
+        foreach ($meta->data->engagement as $engagementtype => $engagementnumber) {
+            $attentiontotals[$engagementtype] = ($attentiontotals[$engagementtype] ?? 0) + ($engagementnumber ? 1 : 0);
```

Comment:
Could this be achieved more elegantly in an array_reduce?

Reply:
How did you think this would look like with an array reduce? Because this is an associative array I'm looping through so the key isn't normally available in an array reduce. Even though this is a nested foreach loop, I was hoping the comment would make what it's trying to do understandable enough.

Reply:
I just tried it on both levels, does either make more sense to you?
```
    $attentiontotals = array_reduce($attendees, function($acc, $attendee){
        $meta = json_decode($attendee->meta);
        $engagement = $meta->data->engagement;
        foreach ($engagement as $engagementtype => $engagementnumber) {
            $acc[$engagementtype] = ($acc[$engagementtype] ?? 0) + ($engagementnumber ? 1 : 0);
        }
        return $acc;
    }, $attentiontotals);

    foreach ($attendees as &$attendee) {
        $meta = json_decode($attendee->meta);
        $engagement = (array)$meta->data->engagement;
        $attentiontotals = array_reduce(array_keys($engagement), function($acc, $key) use ($engagement) {
            $acc[$key] = ($acc[$key] ?? 0) + ($engagement[$key] ? 1 : 0);
            return $acc;
        }, $attentiontotals);
    }


```

Reply:
The foreach is fine, appreciate the examples to see. was Just wondering as the logic is complex, but its complex either way.